File: testing/tests_js/test_011_view_transitions_frontend.py
"""Node-backed checks for frontend view transition behavior."""

import textwrap

def run_transition_check(run_node, assertion: str):
    script = f"""
const fs = require("node:fs");
const vm = require("node:vm");

const capturedErrors = [];
const events = [];
let transitionStarts = 0;
let finishTransition = null;

const context = {{
  capturedErrors,
  clearTimeout,
  console,
  crypto,
  document: {{
    startViewTransition(callback) {{
      transitionStarts += 1;
      const updateCallbackDone = Promise.resolve().then(callback);
      const finished = new Promise((resolve) => {{
        finishTransition = resolve;
      }});
      return {{
        finished,
        ready: Promise.resolve(),
        updateCallbackDone,
      }};
    }},
  }},
  events,
  get transitionStarts() {{
    return transitionStarts;
  }},
  finishTransition() {{
    finishTransition?.();
  }},
  MutationObserver: class {{
    disconnect() {{}}
    observe() {{}}
  }},
  queueMicrotask,
  setTimeout,
}};

vm.createContext(context);
let source = fs.readFileSync("src/script/shared/utilities.mjs", "utf8");
source = source.replace(
  'import {{ captureError, isSkippedViewTransitionError }} from "./errors";',
  `
const captureError = (...args) => capturedErrors.push(args);
const isSkippedViewTransitionError = () => false;
`,
);
source = source.replaceAll("export const ", "const ");
source = source.replace("export function waitForAttribute", "function waitForAttribute");
source += "\\nglobalThis.withTransition = withTransition;";
vm.runInContext(source, context);
const withTransition = context.withTransition;

(async () => {{
{textwrap.indent(assertion, "  ")}
}})().catch((error) => {{
  console.error(error);
  process.exit(1);
}});
"""

    run_node(script)


# @features view-transition
# @dimensions nested-callback error-reporting
def test_nested_transition_joins_active_transition_without_error_report(run_node):
    run_transition_check(
        run_node,
        """
const result = await withTransition(() => {
  events.push("outer-start");
  void withTransition(() => {
    events.push("inner");
  });
  events.push("outer-end");
});

if (result !== true) {
  throw new Error(`Expected outer transition to complete, got ${result}`);
}
if (context.transitionStarts !== 1) {
  throw new Error(`Expected one browser transition, got ${context.transitionStarts}`);
}
if (capturedErrors.length !== 0) {
  throw new Error(`Nested transition reported ${capturedErrors.length} errors`);
}
if (events.join(",") !== "outer-start,inner,outer-end") {
  throw new Error(`Unexpected event order: ${events.join(",")}`);
}
""",
    )


# @features view-transition
# @dimensions update-completion animation-lifecycle queueing
def test_transition_resolves_after_update_without_waiting_for_animation(run_node):
    run_transition_check(
        run_node,
        """
let resolved = false;
const pending = withTransition(() => events.push("commit")).then(() => {
  resolved = true;
  events.push("resolved");
});
await pending;

if (!resolved || events.join(",") !== "commit,resolved") {
  throw new Error(`Transition waited for animation: ${events.join(",")}`);
}
context.finishTransition();
""",
    )


# @features view-transition
# @dimensions coalescing exact-once
def test_same_turn_commits_share_one_transition_and_run_once(run_node):
    run_transition_check(
        run_node,
        """
const first = withTransition(() => events.push("first"));
const second = withTransition(() => events.push("second"));
const results = await Promise.all([first, second]);

if (context.transitionStarts !== 1) {
  throw new Error(`Expected one coalesced transition, got ${context.transitionStarts}`);
}
if (events.join(",") !== "first,second") {
  throw new Error(`Unexpected commits: ${events.join(",")}`);
}
if (results.some((result) => result !== true)) {
  throw new Error(`Unexpected results: ${results.join(",")}`);
}
context.finishTransition();
""",
    )


# @features view-transition
# @dimensions ready-rejection exact-once
def test_ready_rejection_does_not_replay_commit(run_node):
    script = r'''
const fs = require("node:fs");
const vm = require("node:vm");
const capturedErrors = [];
let commits = 0;
const skipped = new Error("Transition was skipped");
const context = {
  capturedErrors,
  clearTimeout,
  console,
  crypto,
  document: {
    startViewTransition(callback) {
      const updateCallbackDone = Promise.resolve().then(callback);
      return {
        updateCallbackDone,
        ready: Promise.reject(skipped),
        finished: Promise.resolve(),
      };
    },
  },
  MutationObserver: class { disconnect() {} observe() {} },
  queueMicrotask,
  setTimeout,
};
vm.createContext(context);
let source = fs.readFileSync("src/script/shared/utilities.mjs", "utf8");
source = source.replace(
  'import { captureError, isSkippedViewTransitionError } from "./errors";',
  `
const captureError = (...args) => capturedErrors.push(args);
const isSkippedViewTransitionError = (error) => /skipped/i.test(error?.message || "");
`,
);
source = source.replaceAll("export const ", "const ");
source = source.replace("export function waitForAttribute", "function waitForAttribute");
source += "\nglobalThis.withTransition = withTransition;";
vm.runInContext(source, context);

(async () => {
  const result = await context.withTransition(() => { commits += 1; });
  await Promise.resolve();
  if (!result || commits !== 1) throw new Error(`Commit count: ${commits}`);
  if (capturedErrors.length !== 0) {
    throw new Error(`Skipped transition reported ${capturedErrors.length} errors`);
  }
})().catch((error) => {
  console.error(error);
  process.exit(1);
});
'''
    run_node(script)

File: testing/tests_js/test_011_view_transitions_frontend.mjs
import assert from "node:assert/strict";
import { test } from "node:test";
import esmock from "esmock";
import { createBrowser } from "../utility/js/environment.mjs";

async function setupTransitions(t, { skipped = () => false } = {}) {
	createBrowser(t);
	const capturedErrors = [];
	const captureError = t.mock.fn((...args) => capturedErrors.push(args));
	const { withTransition } = await esmock.strict(
		"../../src/script/shared/transitions.mjs",
		{
			"../../src/script/shared/errors.mjs": {
				captureError,
				isSkippedViewTransitionError: skipped,
			},
		},
	);
	return { capturedErrors, withTransition };
}

function installTransitionApi(t, implementation) {
	const transitionDocument = document;
	Object.defineProperty(transitionDocument, "startViewTransition", {
		configurable: true,
		value: implementation,
		writable: true,
	});
	t.after(() => {
		delete transitionDocument.startViewTransition;
	});
}

function createPendingAnimation(callback) {
	let finish;
	const updateCallbackDone = Promise.resolve().then(callback);
	const finished = new Promise((resolve) => {
		finish = resolve;
	});
	return {
		finish,
		transition: {
			finished,
			ready: Promise.resolve(),
			updateCallbackDone,
		},
	};
}

/** @matrix view-transition : error-reporting nested-callback */
test("test_nested_transition_joins_active_transition_without_error_report", async (t) => {
	const { capturedErrors, withTransition } = await setupTransitions(t);
	const events = [];
	let transitionStarts = 0;
	let finishTransition;
	installTransitionApi(t, (callback) => {
		transitionStarts += 1;
		const pending = createPendingAnimation(callback);
		finishTransition = pending.finish;
		return pending.transition;
	});

	const result = await withTransition(() => {
		events.push("outer-start");
		void withTransition(() => events.push("inner"));
		events.push("outer-end");
	});

	assert.equal(result, true);
	assert.equal(transitionStarts, 1);
	assert.equal(capturedErrors.length, 0);
	assert.deepEqual(events, ["outer-start", "inner", "outer-end"]);
	finishTransition();
});

/** @matrix view-transition : animation-lifecycle queueing update-completion */
test("test_transition_resolves_after_update_without_waiting_for_animation", async (t) => {
	const { withTransition } = await setupTransitions(t);
	const events = [];
	let finishTransition;
	installTransitionApi(t, (callback) => {
		const pending = createPendingAnimation(callback);
		finishTransition = pending.finish;
		return pending.transition;
	});

	let resolved = false;
	await withTransition(() => events.push("commit")).then(() => {
		resolved = true;
		events.push("resolved");
	});

	assert.equal(resolved, true);
	assert.deepEqual(events, ["commit", "resolved"]);
	finishTransition();
});

/** @matrix view-transition : coalescing exact-once */
test("test_same_turn_commits_share_one_transition_and_run_once", async (t) => {
	const { withTransition } = await setupTransitions(t);
	const events = [];
	let transitionStarts = 0;
	let finishTransition;
	installTransitionApi(t, (callback) => {
		transitionStarts += 1;
		const pending = createPendingAnimation(callback);
		finishTransition = pending.finish;
		return pending.transition;
	});

	const results = await Promise.all([
		withTransition(() => events.push("first")),
		withTransition(() => events.push("second")),
	]);

	assert.equal(transitionStarts, 1);
	assert.deepEqual(events, ["first", "second"]);
	assert.deepEqual(results, [true, true]);
	finishTransition();
});

/** @matrix view-transition : exact-once ready-rejection */
test("test_ready_rejection_does_not_replay_commit", async (t) => {
	const skipped = new Error("Transition was skipped");
	const { capturedErrors, withTransition } = await setupTransitions(t, {
		skipped: (error) => /skipped/i.test(error?.message || ""),
	});
	installTransitionApi(t, (callback) => ({
		finished: Promise.resolve(),
		ready: Promise.reject(skipped),
		updateCallbackDone: Promise.resolve().then(callback),
	}));
	let commits = 0;

	const result = await withTransition(() => {
		commits += 1;
	});
	await Promise.resolve();

	assert.equal(result, true);
	assert.equal(commits, 1);
	assert.equal(capturedErrors.length, 0);
});
